homeassistant/components/life360/device_tracker.py

```diff
@@ -0,0 +1,624 @@
+    vol.Optional(CONF_MEMBERS): _EXCL_INCL_SLUG_LIST,
+    vol.Optional(CONF_PLACES): _EXCL_INCL_LOWER_STRING_LIST,
+    vol.Optional(CONF_PREFIX, default=DEFAULT_PREFIX):
+        vol.All(vol.Any(None, cv.string), _prefix),
+    vol.Optional(CONF_SHOW_AS_STATE, default=[]): vol.All(
+        cv.ensure_list, [vol.In(SHOW_AS_STATE_OPTS)]),
+    vol.Optional(CONF_WARNING_THRESHOLD): _THRESHOLD,
+})
+
+PLATFORM_SCHEMA = vol.All(LIFE360_SCHEMA, _thresholds)
+
+
+def _include_name(filter_dict, name):
+    if not name:
+        return False
+    if not filter_dict:
+        return True
+    name = name.lower()
+    if filter_dict['include']:
+        return name in filter_dict['list']
+    return name not in filter_dict['list']
+
+
+def _exc_msg(exc):
+    return '{}: {}'.format(exc.__class__.__name__, str(exc))
+
+
+def _setup_zone_updating(hass, config, api):
+    from life360 import Life360Error
+
+    add_zones = config.get(CONF_ADD_ZONES)
+    if not add_zones:
+        return
+
+    Place = namedtuple(
+        'Place', [ATTR_NAME, ATTR_LATITUDE, ATTR_LONGITUDE, ATTR_RADIUS])
+
+    def get_places(api, circles_filter, places_filter):
+        errs = 0
+        while True:
+            places = set()
+            try:
+                for circle in api.get_circles():
+                    circle_id = circle['id']
+                    if not (circle_id
+                            and _include_name(circles_filter, circle['name'])):
+                        continue
+                    for place in api.get_circle_places(circle_id):
+                        place_name = place['name']
+                        if not _include_name(places_filter, place_name):
+                            continue
+                        places.add(Place(place_name,
+                                         float(place['latitude']),
+                                         float(place['longitude']),
+                                         float(place['radius'])))
+            except (Life360Error, KeyError, TypeError, ValueError) as exc:
+                errs += 1
+                if errs >= 3:
+                    _LOGGER.error('get_places: %s', _exc_msg(exc))
+                    return None
+            else:
+                return places
+
+    def zone_from_place(place, entity_id=None):
+        zone = Zone(hass, *place, None, DEFAULT_PASSIVE)
+        zone.entity_id = (
+            entity_id or
+            generate_entity_id(ZN_ENTITY_ID_FORMAT, place.name, None, hass))
+        zone.schedule_update_ha_state()
+        return zone
+
+    def log_places(msg, places):
+        plural = 's' if len(places) > 1 else ''
+        _LOGGER.debug(
+            '%s zone%s from Place%s: %s',
+            msg, plural, plural,
+            '; '.join(['{}: {}, {}, {}'.format(*place) for place
+                       in sorted(places, key=lambda x: x.name.lower())]))
+
+    def zones_from_places(api, circles_filter, places_filter, home_place_name,
```

Comment:
The zone handling needs a core review, I think. Is this something we want platforms to do?

@home-assistant/core

Reply:
It may worth to open an architecture issue. Right now, user seems don't have way to create a zone through services or any api call, only way is going through config_flow, and no way to edit/delete them.

I think zone need rewrite, I propose moving zone into the core, like *_registry, not a component.

Reply:
FWIW, this will only create zones during the current session, just like owntracks does. (They are therefore effectively not "permanent", at least as far as HA is concerned.) And if the Life360 Place is removed, the corresponding zone is automatically removed, too. So these are really, to some extent, separate from zones created via configuration.yaml or config_flow, although of course, unique entity_id's will always be created.

I can certainly understand a need to improve zone handling overall. But this works today (and has for quite some time.) Is it ok to use this current implementation and then come back and adjust as needed if/when the zone infrastructure is changed as a separate task? Just a thought...

Reply:
Creating zones is fine. 

It would be better if we would expose an appropriate API from the zone integration instead of having integrations dive into the Zone integration and take it's entity. I guess in a "perfect" world, the zone integration would have platforms and life360 and OwnTracks would be two of them.

For this PR, I suggest to add a `async_create_zone(hass, name, lat, long, radius)` to the Zone integration and call that from Life360.

Reply:
So a few issues/questions that now come to mind:

1. Should `async_create_zone` return the `Zone` object? Or maybe the entity_id?
1. Should `async_create_zone` keep track of the created `Zone` object, and if so, where?
1. Should there also be an `async_remove_zone` function added to remove a zone that was created via the first function? And, if so, what should it take as a parameter -- the entity_id, name, `Zone` object returned by the first function, ...?
1. What is the best way, assuming the questions above have been answered/addressed, to update `zone.home`?

Regarding this last point, I now realize that I'm probably not handling the updating of `zone.home` as well as I should. First, I'm not using the home icon. Second, and possibly more important, I'm creating a `Zone` object for home, but there already is one, so now there would be two. I guess it's "worked" up until now because zone entities on their own never update after they've been created, and to the rest of the system it only matters what's in the state machine. But if we're going to do this in a reasonable manner, maybe updating `zone.home` should actually update the existing `Zone` object in some way.

If we all don't have time right now to work through all these questions (so as to come up with an appropriate enhancement to the zone integration), is it ok to go with what I have now, assuming that last point above isn't too bad of a situation. (I can probably address the icon issue fairly easily.)

Reply:
We should not return the zone object. Entity ID could be ok. 

`async_create_zone` to be able to track and re-create after reboot is what https://github.com/home-assistant/architecture/issues/204 is about. If you are ready to take that one on next… please 😉 

I think that your questions are very specific, and we first need to decide on how zones should work: I think that zones should work as any entity platform, so that integrations own their own zone entities and can create/delete etc as they see fit. I think that the Home zone should be special and cannot be set by integrations.

Reply:
I'll try to work through the details and propose a working solution that fits into the current zone implementation and handles the life360 and owntracks use cases.

Reply:
@balloob didn't realize you were responding at the same time! Lol

Reply:
@balloob see latest commit for a proposed implementation. I did some sanity testing with life360, but I'm not set up to test owntracks. I still need to do more extensive testing (hence the "WIP") with life360, but I thought I'd commit what I have so far to get your feedback. BTW, I also rebased, which is the reason for the force push.

Reply:
I should also mention I think it's important to be able to change the location & size of `zone.home`. This is one of the main reasons I added the ability to create/update HA zones from Life360 Places in the first place.

Reply:
Working on cleaning up tests. Since zone component used `hass.data[DOMAIN]` as a dictionary with (effectively) the object_id's (of configured zones) as the key I'm changing it back to that format (i.e., object_id's as keys instead of entity_id's) since it's not 100% clear to me if anything else in the system depended on that. It also means I don't have to "fix" the corresponding tests. I'll be force pushing soon with these changes.

Reply:
Don't bother reviewing this just yet. I've given it more thought and I want to tweak it a bit more.

Reply:
Sorry for the delay. I wanted to make sure not to change any of the original behavior. I think it's ready for a review. Also let me know if I need to add any tests.

Reply:
Home zone is created based off the core configuration. We can't have integrations start overriding it, because it won't be clear who "owns" the data and there is no longer a single source of truth.

Reply:
Sorry, but I don't understand this.

First, as I already mentioned, the need came from practical experience. If HA and Life360 have different definitions of home then they can, and will, disagree. This has been noticed by the users of my existing custom component and has caused confusion.

Second, to avoid that bad situation the two systems need to come into agreement as to where home is. Either this needs to be done manually, which is error prone and not as easy as it might seem (HA defines via GPS coordinates, whereas Life360 does it via address or a map), or this can be done automatically for the user. At least making HA agree with Life360 can be done automatically. (See [this](https://deploy-preview-9302--home-assistant-docs.netlify.com/components/life360/) on the proposed doc page for more details.)

Third, who's to say whether HA's or Life360's definition of home is correct? I think this should be up to the user. Updating HA's home zone from Life360 info is a configurable feature -- the user can decide whether that's what they want or not.

And lastly, even in HA there is no "single source of truth", because it can come from either configuration.yaml or from an IP address lookup. And even if it comes from the config, there's two choices: under homeassistant or zone. So that's three different places it can come from in HA.

The bottom line to me is, it should be up to the user, and having a way to automatically keep the two systems in sync is better than forcing a user to manually copy data from one system to the other.

Reply:
@balloob so how can we make some progress on this? Maybe I should back out the changes to the zone and owntracks components and just put the life360 code back the way it was, which as I've said, has been not only working well in my custom component that many, many people use, but it also solves the real problems I've outlined. Then you can take your time to work out what you want to do with zones, and then this, as well as owntracks, etc., can be updated accordingly. But in the meantime I can retire my custom component and people can start using a standard component.